// @verdaccio/core/build/plugin-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Plugin = void 0;
/**
 * The base plugin class, set of utilities for developing
 * plugins.
 * @alpha
 * */
class Plugin {
  static version = 1;
  constructor(config, options) {
    this.version = Plugin.version;
    this.config = config;
    this.options = options;
  }
  getVersion() {
    return this.version;
  }
}

/**
 * This function allow add additional middleware to the application.
 *
 *  ```ts
 *  import express, { Request, Response } from 'express';
 * 
 *  class Middleware extends Plugin {
 *    // instances of auth and storage are injected
 *    register_middlewares(app, auth, storage) {
 *      const router = express.Router();
 *      router.post('/my-endpoint', (req: Request, res: Response): void => {
        res.status(200).end();
      });
 *    }
 *  }
 * 
 *  
 *  const [plugin] = await asyncLoadPlugin(...);
 *  plugin.register_middlewares(app, auth, storage);
 *  ```
 */

/**
 * dasdsa
 */
exports.Plugin = Plugin;
