// @verdaccio/signature/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SignOptionsSignature", {
  enumerable: true,
  get: function () {
    return _jwtToken.SignOptionsSignature;
  }
});
Object.defineProperty(exports, "aesDecrypt", {
  enumerable: true,
  get: function () {
    return _signature.aesDecrypt;
  }
});
Object.defineProperty(exports, "aesDecryptDeprecated", {
  enumerable: true,
  get: function () {
    return _legacySignature.aesDecryptDeprecated;
  }
});
Object.defineProperty(exports, "aesEncrypt", {
  enumerable: true,
  get: function () {
    return _signature.aesEncrypt;
  }
});
Object.defineProperty(exports, "aesEncryptDeprecated", {
  enumerable: true,
  get: function () {
    return _legacySignature.aesEncryptDeprecated;
  }
});
Object.defineProperty(exports, "generateRandomSecretKeyDeprecated", {
  enumerable: true,
  get: function () {
    return _legacySignature.generateRandomSecretKeyDeprecated;
  }
});
Object.defineProperty(exports, "parseBasicPayload", {
  enumerable: true,
  get: function () {
    return _token.parseBasicPayload;
  }
});
Object.defineProperty(exports, "signPayload", {
  enumerable: true,
  get: function () {
    return _jwtToken.signPayload;
  }
});
exports.utils = exports.types = void 0;
Object.defineProperty(exports, "verifyPayload", {
  enumerable: true,
  get: function () {
    return _jwtToken.verifyPayload;
  }
});
var _legacySignature = require("./legacy-signature");
var _signature = require("./signature");
var _jwtToken = require("./jwt-token");
var _utils = _interopRequireWildcard(require("./utils"));
exports.utils = _utils;
var _types = _interopRequireWildcard(require("./types"));
exports.types = _types;
var _token = require("./token");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
