// verdaccio/build/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "API_ERROR", {
  enumerable: true,
  get: function () {
    return _core.API_ERROR;
  }
});
Object.defineProperty(exports, "API_MESSAGE", {
  enumerable: true,
  get: function () {
    return _core.API_MESSAGE;
  }
});
Object.defineProperty(exports, "APP_ERROR", {
  enumerable: true,
  get: function () {
    return _core.APP_ERROR;
  }
});
Object.defineProperty(exports, "CHARACTER_ENCODING", {
  enumerable: true,
  get: function () {
    return _core.CHARACTER_ENCODING;
  }
});
exports.ERROR_CODE = exports.DIST_TAGS = exports.DEFAULT_USER = exports.DEFAULT_UPLINK = exports.DEFAULT_REGISTRY = exports.DEFAULT_PROTOCOL = exports.DEFAULT_PORT = exports.DEFAULT_NO_README = exports.DEFAULT_MIN_LIMIT_PASSWORD = exports.DEFAULT_DOMAIN = void 0;
Object.defineProperty(exports, "HEADERS", {
  enumerable: true,
  get: function () {
    return _core.HEADERS;
  }
});
Object.defineProperty(exports, "HEADER_TYPE", {
  enumerable: true,
  get: function () {
    return _core.HEADER_TYPE;
  }
});
Object.defineProperty(exports, "HTTP_STATUS", {
  enumerable: true,
  get: function () {
    return _core.HTTP_STATUS;
  }
});
exports.MODULE_NOT_FOUND = exports.LATEST = void 0;
Object.defineProperty(exports, "PACKAGE_ACCESS", {
  enumerable: true,
  get: function () {
    return _utils.PACKAGE_ACCESS;
  }
});
Object.defineProperty(exports, "ROLES", {
  enumerable: true,
  get: function () {
    return _utils.ROLES;
  }
});
exports.STORAGE = void 0;
Object.defineProperty(exports, "SUPPORT_ERRORS", {
  enumerable: true,
  get: function () {
    return _core.SUPPORT_ERRORS;
  }
});
exports.TIME_EXPIRATION_24H = exports.TIME_EXPIRATION_1H = void 0;
Object.defineProperty(exports, "TOKEN_BASIC", {
  enumerable: true,
  get: function () {
    return _core.TOKEN_BASIC;
  }
});
Object.defineProperty(exports, "TOKEN_BEARER", {
  enumerable: true,
  get: function () {
    return _core.TOKEN_BEARER;
  }
});
exports.keyPem = exports.csrPem = exports.certPem = exports.WEB_TITLE = exports.USERS = void 0;
var _core = require("@verdaccio/core");
var _utils = require("@verdaccio/utils");
const DEFAULT_PORT = exports.DEFAULT_PORT = '4873';
const DEFAULT_PROTOCOL = exports.DEFAULT_PROTOCOL = 'http';
const DEFAULT_DOMAIN = exports.DEFAULT_DOMAIN = 'localhost';
const TIME_EXPIRATION_24H = exports.TIME_EXPIRATION_24H = '24h';
const TIME_EXPIRATION_1H = exports.TIME_EXPIRATION_1H = '1h';
const DIST_TAGS = exports.DIST_TAGS = 'dist-tags';
const LATEST = exports.LATEST = 'latest';
const USERS = exports.USERS = 'users';
const DEFAULT_MIN_LIMIT_PASSWORD = exports.DEFAULT_MIN_LIMIT_PASSWORD = 3;
const DEFAULT_USER = exports.DEFAULT_USER = 'Anonymous';
const keyPem = exports.keyPem = 'verdaccio-key.pem';
const certPem = exports.certPem = 'verdaccio-cert.pem';
const csrPem = exports.csrPem = 'verdaccio-csr.pem';
const ERROR_CODE = exports.ERROR_CODE = {
  token_required: 'token is required'
};
const DEFAULT_REGISTRY = exports.DEFAULT_REGISTRY = 'https://registry.npmjs.org';
const DEFAULT_UPLINK = exports.DEFAULT_UPLINK = 'npmjs';
const DEFAULT_NO_README = exports.DEFAULT_NO_README = 'ERROR: No README data found!';
const MODULE_NOT_FOUND = exports.MODULE_NOT_FOUND = 'MODULE_NOT_FOUND';
const WEB_TITLE = exports.WEB_TITLE = 'Verdaccio';
const STORAGE = exports.STORAGE = {
  PACKAGE_FILE_NAME: 'package.json',
  FILE_EXIST_ERROR: 'EEXISTS',
  NO_SUCH_FILE_ERROR: 'ENOENT',
  DEFAULT_REVISION: '0-0000000000000000'
};
