// @verdaccio/middleware/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  match: true,
  setSecurityWebHeaders: true,
  validateName: true,
  validatePackage: true,
  media: true,
  makeURLrelative: true,
  encodeScopePackage: true,
  expectJson: true,
  antiLoop: true,
  final: true,
  allow: true,
  rateLimit: true,
  userAgent: true,
  webMiddleware: true,
  renderWebMiddleware: true,
  errorReportingMiddleware: true,
  handleError: true,
  log: true,
  LOG_STATUS_MESSAGE: true,
  LOG_VERDACCIO_BYTES: true,
  LOG_VERDACCIO_ERROR: true
};
Object.defineProperty(exports, "LOG_STATUS_MESSAGE", {
  enumerable: true,
  get: function () {
    return _log.LOG_STATUS_MESSAGE;
  }
});
Object.defineProperty(exports, "LOG_VERDACCIO_BYTES", {
  enumerable: true,
  get: function () {
    return _log.LOG_VERDACCIO_BYTES;
  }
});
Object.defineProperty(exports, "LOG_VERDACCIO_ERROR", {
  enumerable: true,
  get: function () {
    return _log.LOG_VERDACCIO_ERROR;
  }
});
Object.defineProperty(exports, "allow", {
  enumerable: true,
  get: function () {
    return _allow.allow;
  }
});
Object.defineProperty(exports, "antiLoop", {
  enumerable: true,
  get: function () {
    return _antiLoop.antiLoop;
  }
});
Object.defineProperty(exports, "encodeScopePackage", {
  enumerable: true,
  get: function () {
    return _encodePkg.encodeScopePackage;
  }
});
Object.defineProperty(exports, "errorReportingMiddleware", {
  enumerable: true,
  get: function () {
    return _error.errorReportingMiddleware;
  }
});
Object.defineProperty(exports, "expectJson", {
  enumerable: true,
  get: function () {
    return _json.expectJson;
  }
});
Object.defineProperty(exports, "final", {
  enumerable: true,
  get: function () {
    return _final.final;
  }
});
Object.defineProperty(exports, "handleError", {
  enumerable: true,
  get: function () {
    return _error.handleError;
  }
});
Object.defineProperty(exports, "log", {
  enumerable: true,
  get: function () {
    return _log.log;
  }
});
Object.defineProperty(exports, "makeURLrelative", {
  enumerable: true,
  get: function () {
    return _makeUrlRelative.makeURLrelative;
  }
});
Object.defineProperty(exports, "match", {
  enumerable: true,
  get: function () {
    return _match.match;
  }
});
Object.defineProperty(exports, "media", {
  enumerable: true,
  get: function () {
    return _media.media;
  }
});
Object.defineProperty(exports, "rateLimit", {
  enumerable: true,
  get: function () {
    return _rateLimit.rateLimit;
  }
});
Object.defineProperty(exports, "renderWebMiddleware", {
  enumerable: true,
  get: function () {
    return _web.renderWebMiddleware;
  }
});
Object.defineProperty(exports, "setSecurityWebHeaders", {
  enumerable: true,
  get: function () {
    return _securityHeaders.setSecurityWebHeaders;
  }
});
Object.defineProperty(exports, "userAgent", {
  enumerable: true,
  get: function () {
    return _userAgent.userAgent;
  }
});
Object.defineProperty(exports, "validateName", {
  enumerable: true,
  get: function () {
    return _validation.validateName;
  }
});
Object.defineProperty(exports, "validatePackage", {
  enumerable: true,
  get: function () {
    return _validation.validatePackage;
  }
});
Object.defineProperty(exports, "webMiddleware", {
  enumerable: true,
  get: function () {
    return _web.webMiddleware;
  }
});
var _match = require("./middlewares/match");
var _securityHeaders = require("./middlewares/security-headers");
var _validation = require("./middlewares/validation");
var _media = require("./middlewares/media");
var _makeUrlRelative = require("./middlewares/make-url-relative");
var _encodePkg = require("./middlewares/encode-pkg");
var _json = require("./middlewares/json");
var _antiLoop = require("./middlewares/antiLoop");
var _final = require("./middlewares/final");
var _allow = require("./middlewares/allow");
var _rateLimit = require("./middlewares/rate-limit");
var _userAgent = require("./middlewares/user-agent");
var _web = require("./middlewares/web");
var _error = require("./middlewares/error");
var _log = require("./middlewares/log");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _api_urls = require("./middlewares/api_urls");
Object.keys(_api_urls).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _api_urls[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _api_urls[key];
    }
  });
});
var _webUrls = require("./middlewares/web/web-urls");
Object.keys(_webUrls).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _webUrls[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _webUrls[key];
    }
  });
});
